// @astrojs/partytown/dist/index.js
import { partytownSnippet } from "@builder.io/partytown/integration";
import { copyLibFiles, libDirPath } from "@builder.io/partytown/utils";
import * as fs from "fs";
import { createRequire } from "module";
import path from "path";
import { fileURLToPath } from "url";
import sirv from "./sirv.js";
const resolve = createRequire(import.meta.url).resolve;
function appendForwardSlash(str) {
  return str.endsWith("/") ? str : str + "/";
}
function createPlugin(options) {
  let config;
  let partytownSnippetHtml;
  const partytownEntrypoint = resolve("@builder.io/partytown/package.json");
  const partytownLibDirectory = path.resolve(partytownEntrypoint, "../lib");
  return {
    name: "@astrojs/partytown",
    hooks: {
      "astro:config:setup": ({ config: _config, command, injectScript }) => {
        var _a;
        const lib = `${appendForwardSlash(_config.base)}~partytown/`;
        const partytownConfig = {
          ...options == null ? void 0 : options.config,
          lib,
          debug: ((_a = options == null ? void 0 : options.config) == null ? void 0 : _a.debug) ?? command === "dev"
        };
        partytownSnippetHtml = partytownSnippet(partytownConfig);
        injectScript("head-inline", partytownSnippetHtml);
      },
      "astro:config:done": ({ config: _config }) => {
        config = _config;
      },
      "astro:server:setup": ({ server }) => {
        const lib = `/~partytown/`;
        server.middlewares.use(
          sirv(partytownLibDirectory, {
            mount: lib,
            dev: true,
            etag: true,
            extensions: []
          })
        );
      },
      "astro:build:done": async ({ dir }) => {
        var _a;
        await copyLibFiles(fileURLToPath(new URL("~partytown", dir)), {
          debugDir: ((_a = options == null ? void 0 : options.config) == null ? void 0 : _a.debug) ?? false
        });
      },
      "astro:build:ssr": async ({ manifest }) => {
        const dirpath = libDirPath({ debugDir: false });
        const files = await fs.promises.readdir(dirpath);
        for (const file of files) {
          if (file === "debug")
            continue;
          manifest.assets.push(`/~partytown/${file}`);
        }
      }
    }
  };
}
export {
  createPlugin as default
};
